internal/outpost/ldap/search/memory/memory.go

```diff
@@ -102,23 +103,23 @@ func (ms *MemorySearcher) Search(req *search.Request) (ldap.ServerSearchResult,
 	var groups []*group.LDAPGroup
 
 	if needUsers {
-		if flags.CanSearch {
+		if flag.CanSearch {
 			users = &ms.users
 		} else {
 			u := make([]api.User, 1)
-			if flags.UserInfo == nil {
+			if flag.UserInfo == nil {
 				for i, u := range ms.users {
-					if u.Pk == flags.UserPk {
-						flags.UserInfo = &ms.users[i]
+					if u.Pk == flag.UserPk {
+						flag.UserInfo = &ms.users[i]
 					}
 				}
-
-				if flags.UserInfo == nil {
-					req.Log().WithField("pk", flags.UserPk).Warning("User with pk is not in local cache")
+				if flag.UserInfo == nil {
+					req.Log().WithField("pk", flag.UserPk).Warning("User with pk is not in local cache")
 					err = fmt.Errorf("failed to get userinfo")
 				}
-			} else {
```

Comment:
I think this else could probably stay as we check for userinfo being nil, so if that's not the case it has to be non-nil

Reply:
No, because if we set the userinfo in the first branch, the else-branch is not executed, leading to the described bug.

Alternative would be to set in both the `u[0]` variable

Reply:
Good point, the more I read through this code the more I'm realising it needs a refactor

Reply:
I needed to read it also a few times. Maybe extract a function from it, which returns either (userinfo, error)?

Reply:
I was thinking this, but yeah there's room for more optimisation

```go
	if needUsers {
		if flag.CanSearch {
			users = &ms.users
		} else if flag.UserInfo == nil {
			for i, u := range ms.users {
				if u.Pk == flag.UserPk {
					flag.UserInfo = &ms.users[i]
					users = &[]api.User{
						ms.users[i],
					}
				}
			}
			if flag.UserInfo == nil {
				req.Log().WithField("pk", flag.UserPk).Warning("User with pk is not in local cache")
				err = fmt.Errorf("failed to get userinfo")
			}
		} else {
			users = &[]api.User{
				*flag.UserInfo,
			}
		}
	}

```

Reply:
Good point, allthough actually thinking about it some more, I'll revert most of the refactoring, I'm a lot more comfortable with changing that much when we have some actual tests